--- a/arch/arm64/boot/dts/rockchip/rk3399-clockworkpi-a06.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-clockworkpi-a06.dts
@@ -20,7 +20,7 @@
 	compatible = "clockwork,devterm-a06", "rockchip,rk3399";
 
 	chosen {
-		stdout-path = "serial2:1500000n8";
+		stdout-path = "serial2:115200n8";
 	};
 
 	aliases {
@@ -357,7 +357,7 @@
 		compatible = "rockchip,rk808";
 		reg = <0x1b>;
 		interrupt-parent = <&gpio1>;
-		interrupts = <21 IRQ_TYPE_LEVEL_LOW>;
+		interrupts = <21 IRQ_TYPE_LEVEL_HIGH>;
 		#clock-cells = <1>;
 		clock-output-names = "xin32k", "rk808-clkout2";
 		pinctrl-names = "default";
@@ -742,7 +742,7 @@
 	rockchip,i2s-broken-burst-len;
 	rockchip,playback-channels = <8>;
 	rockchip,capture-channels = <8>;
-	pinctrl-0 = <&i2s0_2ch_bus>;
+	pinctrl-0 = <&i2s0_8ch_bus>;
 	status = "okay";
 };
 
