--- a/arch/arm64/boot/dts/rockchip/rk3399-clockworkpi-a06.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-clockworkpi-a06.dts
@@ -192,8 +192,12 @@
 
 	battery: battery@0 {
 		compatible = "simple-battery";
+   device-chemistry = "lithium-ion";
 		constant-charge-current-max-microamp = <2100000>;
-		voltage-min-design-microvolt = <3300000>;
+		voltage-min-design-microvolt = <2850000>;
+		voltage-max-design-microvolt = <4200000>;
+		re-charge-voltage-microvolt = <4000000>;
+   charge-full-design-microamp-hours = <3350000>;
 	};
 
 	ocp8178_backlight: backlight@0 {
